src/devices/src/virtio/vsock/device.rs

```diff
@@ -44,54 +52,155 @@ use super::{defs, defs::uapi, EpollConfig};
 const AVAIL_FEATURES: u64 =
     1 << uapi::VIRTIO_F_VERSION_1 as u64 | 1 << uapi::VIRTIO_F_IN_ORDER as u64;
 
-pub struct Vsock<B: VsockBackend> {
+pub struct Vsock<B: VsockBackend + 'static> {
     cid: u64,
-    backend: Option<B>,
+    queues: Vec<VirtQueue>,
+    pub(crate) queue_events: Vec<EventFd>,
+    mem: GuestMemory,
+    pub(crate) backend: B,
     avail_features: u64,
     acked_features: u64,
-    epoll_config: EpollConfig,
-    interrupt_evt: EventFd,
     interrupt_status: Arc<AtomicUsize>,
-    queues: Vec<VirtQueue>,
-    queue_evts: Vec<EventFd>,
+    interrupt_evt: EventFd,
+    // This EventFd is the only one initially registered for a vsock device, and is used to convert
+    // a VirtioDevice::activate call into an EventHandler read event which allows the other events
+    // (queue and backend related) to be registered post virtio device activation. That's
+    // mostly something we wanted to happen for the backend events, to prevent (potentially)
+    // continuous triggers from happening before the device gets activated.
+    pub(crate) activate_evt: EventFd,
```

Comment:
It's a bit trickier here because the backend defines its own characteristics via the `VsockEpollListener::get_polled_evset` method, so this is one way to make sure things are fine no matter what for now. But yeah, the discussion you are proposing makes sense going forward.